File: Objc/TapPayTokenPushExample/TapPayTokenPushExample/TermsViewController.m
//
//  TermsViewController.m
//  TapPayTokenPushExample
//
//  Created by Cherri_TapPay_LukeCho on 2022/8/10.
//

#import "TermsViewController.h"

@interface TermsViewController ()

@property (weak, nonatomic) IBOutlet UIView *termsView;
@property (weak, nonatomic) IBOutlet UIButton *disagreeBtn;
@property (weak, nonatomic) IBOutlet UIButton *agreeBtn;
@property (weak, nonatomic) IBOutlet UIImageView *checkboxImageView;
@property (assign, nonatomic) BOOL isChecked;

@end

@implementation TermsViewController

#pragma mark - Life Cycle

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self configTermsViewWithCheckStatus:_isChecked];
    UITapGestureRecognizer *tapGesture = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(termsViewTouched)];
    [_termsView addGestureRecognizer:tapGesture];
}

#pragma mark - Private Method

- (void)termsViewTouched {
    _isChecked = !_isChecked;
    [self configTermsViewWithCheckStatus:_isChecked];
}

- (void)configTermsViewWithCheckStatus:(BOOL)isCheked {
    if (isCheked) {
        [_checkboxImageView setImage:[UIImage systemImageNamed:@"checkmark.square"]];
        [_agreeBtn setEnabled:true];
    }else {
        [_checkboxImageView setImage:[UIImage systemImageNamed:@"square"]];
        [_agreeBtn setEnabled:false];
    }
}

#pragma mark - IBAction

- (IBAction)disagreePressed:(id)sender {
    [self dismissViewControllerAnimated:false completion:^{
        if ([self.delegate respondsToSelector:@selector(didCancelReadingTerms:)]) {
            [self.delegate didCancelReadingTerms:self];
        }
    }];
}

- (IBAction)agreePressed:(id)sender {
    [self dismissViewControllerAnimated:false completion:^{
        if ([self.delegate respondsToSelector:@selector(didFinishReadingTerms:)]) {
            [self.delegate didFinishReadingTerms:self];
        }
    }];
}


@end

File: Swift/TapPayTokenPushExample/TapPayTokenPushExample/TermsViewController.swift
//
//  TermsViewController.swift
//  TapPayTokenPushExample
//
//  Created by Cherri_TapPay_LukeCho on 2022/8/12.
//

import Foundation
import UIKit

protocol TermsViewControllerDelegate: NSObjectProtocol {
    func didFinishReadingTerms(controller: TermsViewController)
    func didCancelReadingTerms(controller: TermsViewController)
}

class TermsViewController: BaseViewController {
    
    @IBOutlet weak var termsView: UIView!
    @IBOutlet weak var disagreeBtn: UIButton!
    @IBOutlet weak var agreeBtn: UIButton!
    @IBOutlet weak var checkboxImageView: UIImageView!
    var isChecked: Bool = false
    var delegate: TermsViewControllerDelegate?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        configTermsViewWithCheckStatus(isChecked: isChecked)
        let tapGesture = UITapGestureRecognizer.init(target: self, action: #selector(termsViewTouched))
        termsView.addGestureRecognizer(tapGesture)
    }
    
    @objc private func termsViewTouched() {
        isChecked = !isChecked
        configTermsViewWithCheckStatus(isChecked: isChecked)
    }
    
    private func configTermsViewWithCheckStatus(isChecked: Bool) {
        if isChecked {
            checkboxImageView.image = UIImage.init(systemName: "checkmark.square")
            agreeBtn.isEnabled = true
        }else {
            checkboxImageView.image = UIImage.init(systemName: "square")
            agreeBtn.isEnabled = false
        }
    }
    
    @IBAction func disagreePressed(_ sender: Any) {
        self.dismiss(animated: false) {
            self.delegate?.didCancelReadingTerms(controller: self)
        }
    }
    
    @IBAction func agreePressed(_ sender: Any) {
        self.dismiss(animated: false) {
            self.delegate?.didFinishReadingTerms(controller: self)
        }
    }
}
